### PR TITLE
2.2 pingbatcher 1720155

### PR DESCRIPTION
**Pull request by jameinel**
## Description of change

It turns out that $bit operations do not work well with Upsert and Mongo 2.4. I had written the original patch against Mongo 2.6 which is why I thought everything worked, but it seems that things break if you use the 2.4 that we have on Trusty.

## QA steps

The existing test suite would fail if run against Mongo 2.4 which should have caught this, but apparently were weren't actually running there? An easy way to test this is to grab a trusty VM/LXD container, and then 'apt install juju-mongodb' and run the test suite in state/presence.

## Documentation changes

None

## Bug reference

[lp:1720155](https://bugs.launchpad.net/juju/+bug/1720155)

### CODE REVIEW

**Comment by nskaggs** [comment]
We'd planned on getting 2.6 into trusty, but it never happened. It seems we should consider restoring running on mongo2.4 and trusty, or restart the effort to move to mongo 2.6

File: state/presence/pingbatcher.go [hunk 1]
```diff
@@ -298,25 +354,25 @@ func (pb *PingBatcher) flush() error {
...
-			bson.D{
-				{"$set", bson.D{{"slot", slot.Slot}}},
-				{"$bit", fields},
-			},
-		)
+		_, err := pings.UpsertId(docId, update)
 		if err != nil {
 			return errors.Trace(err)
 		}
+		// if changeInfo.Updated == 0 {
```
Comment (wallyworld) [inline state/presence/pingbatcher.go]: remove me?

Comment (jameinel) [reply state/presence/pingbatcher.go]: done

**Comment by jameinel** [comment]
$$merge$$

**Pull request merged**